File: php/src/Advertisement/Application/Errors/AdvertisementAlreadyExistsError.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisement\Application\Errors;

use Demo\App\Common\Result;

final readonly class AdvertisementAlreadyExistsError
{
    public static function build(string $id): Result
    {
        return Result::failure(sprintf('Advertisement with id %s already exists', $id));
    }
}

File: kotlin/src/main/kotlin/advertisement/domain/errors/AdvertisementAlreadyExistsError.kt
package advertisement.domain.errors

import common.domain.DomainError

open class AdvertisementAlreadyExistsError private constructor(errorMessage: String): DomainError(errorMessage) {
    companion object {
        fun withId(value: String): AdvertisementAlreadyExistsError {
            val errorMessage = "Advertisement with id $value already exists"

            return AdvertisementAlreadyExistsError(errorMessage)
        }
    }
}
